### ISSUE TITLE
`GitRepository` pull fails

### ISSUE THREAD

**Issue by pypae** [PrefectHQ/prefect#16241]
## Context
 I’m trying to set up `GitRepository` storage to fetch flow definitions from a GitLab repository and run them with `.serve` as described [in the docs](https://docs.prefect.io/v3/deploy/run-flows-in-local-processes#retrieve-a-flow-from-remote-storage).

## Problem
The storage setup fails to pull the repository correctly. The first `GitRepository.pull_code` is successful, because it clones the repo. All subsequent pulls fail with the following error:
```
ERROR   | prefect.runner.storage.git-repository.data-collection - Failed to pull latest changes with exit code Command '['git', 'pull', 'origin', '--depth', '1']' returned non-zero exit status 128
```

I think the issue is with shallow clones and my setup, because I’ve tried manually running `git clone <repo> --depth 1`  and making some changes, then pulling again with `git pull origin --depth 1`. This also fails with the following exception: `fatal: refusing to merge unrelated histories.` I’ve tried setting `git config --global pull.rebase true`. But that doesn’t help.

How can I configure git to work with the prefect `GitRepository` storage?

### Version info

```Text
Version:             3.1.4
API version:         0.8.4
Python version:      3.10.12
Git commit:          78ee41cb
Built:               Wed, Nov 20, 2024 7:37 PM
OS/Arch:             linux/x86_64
Profile:             ephemeral
Server type:         server
Pydantic version:    2.10.2
```

### Additional context

Here's an excerpt of my code:

`prefect_environment.py`
```python
import os
import subprocess
from pathlib import Path

from prefect.runner.storage import GitRepository


def _get_project_source():
    return Path(__file__).parent.parent.parent


def get_default_config():
    relative_config = "resources/sources/config.yml"
    return _get_project_source() / relative_config


def get_deamon_source():
    # Running on a server.
    repo = os.environ.get("GITLAB_REPO")
    if repo:
        access_token = os.environ.get("GITLAB_ACCESS_TOKEN")
        credentials = None
        if access_token:
            credentials = {"access_token": access_token}
        repo = GitRepository(repo, credentials)
        return repo
    else:
        # Running locally.
        return str(_get_project_source())

def get_git_commit():
    # Check if the directory is a valid git repository
    try:
        commit_hash = subprocess.check_output(
            ["git", "-C", _get_project_source(), "rev-parse", "--short", "HEAD"],
            text=True,
        ).strip()
        return commit_hash
    except subprocess.CalledProcessError:
        print(f"Could not retrieve git commit for directory: {_get_project_source()}")
        return None
```

`prefect_deamon.py`
```python
from datetime import timedelta

from prefect import flow

from data_collection.prefect_environment import get_deamon_source

if __name__ == "__main__":
    deployment = flow.from_source(
        source=get_deamon_source(),
        entrypoint="src/data_collection/prefect_flow.py:download_all",
    )

    # Schedule the deployment to run every day
    deployment.serve(
        name="data-collection-deamon",
        interval=timedelta(hours=24),
    )
```

**Comment by zzstoatzz** [comment]
hi @pypae - thanks for the issue! can you clarify why you've implemented `get_daemon_source`? this should happen for you if you provide something like the following
```python
if __name__ == "__main__":
    deployment = flow.from_source(
        source=repo = GitRepository(os.environ.get("GITLAB_REPO"), {"access_token": access_token}),
        entrypoint="src/data_collection/prefect_flow.py:download_all",
    )

    # Schedule the deployment to run every day
    deployment.serve(
        name="data-collection-deamon",
        interval=timedelta(hours=24),
    )
```

**Comment by pypae** [comment]
Sure, when running locally during development, I don't set `GITLAB_REPO`, so this part of the code pulls in the local flow:

```python
    ...
    else:
        # Running locally.
        return str(_get_project_source())
```

Otherwise I'd have to push my changes to test the source, right?

**Comment by pypae** [comment]
I'm not sure we're on the same page.

To clarify:
I'm running `prefect server start` on a staging server, where I set `GITLAB_REPO` and `GITLAB_ACCESS_TOKEN`. 
I'm then running `python prefect_deamon.py` and then I'm running into the issue with the updates not being pulled after the initial clone. 

The `get_default_config` part is just in place so I can develop locally, I.e. run `prefect server start` locally:
- I edit the `prefect_flow.py` locally without pushing anything yet.
- I then run `python prefect_deamon.py` (without setting any environment variables) and it picks up the local changes to `prefect_flow.py` 

Sorry for the confusion, does this make sense?

**Comment by desertaxle** [comment]
Hey @pypae! Looking at the code for `GitRepository`: https://github.com/PrefectHQ/prefect/blob/09412d17bf80327f0e71f9d88f6f5b4aa9975c96/src/prefect/runner/storage.py#L221-L225
It looks like maybe that log should be a warning instead of an error since we reclone the repo if a pull from origin fails. If your serve process is starting up and running just fine, then I think we can change that log to a warning. If the serve process is failing because of this, then we need to dig in more.

It's hard to tell why cloning is working, but pulling from origin isn't working. My hunch is that your access token might not be available on pull. To check if this is the case, you can turn on debug logging by setting the `PREFECT_LOGGING_LEVEL` environment variable to `DEBUG`. You should be able to see in the logs where Prefect is cloning your repository and then run `git config --get remote.origin.url` to see your origin URL. If it doesn't contain the auth token, then that could be the issue.

**Comment by pypae** [comment]
Thanks for your help!

> It looks like maybe that log should be a warning instead of an error since we reclone the repo if a pull from origin fails.

Sounds good to me. I enabled debug logs and see that after `pull_interval` seconds, the runner successfully pulls the newest commit.

**Comment by pypae** [comment]
Thank you your work by the way, I'm loving `prefect` so far.

**Comment by desertaxle** [comment]
> Sounds good to me. I enabled debug logs and see that after `pull_interval` seconds, the runner successfully pulls the newest commit.

That's good to hear! Is the error only showing when the serve process starts?

**Comment by pypae** [comment]
No, it shows up each time I push a new commit to the repo when it tries to pull it for the first time.

**Comment by jasperjonkhans** [comment]
Is this still available? If yes I would like to give it a shot.
